src/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,174 @@
+#include "software/ai/passing/pass_generator.h"
+
+#include <iomanip>
+
+#include "software/logger/logger.h"
+
+PassGenerator::PassGenerator(const TbotsProto::PassingConfig& passing_config)
+    : optimizer_(optimizer_param_weights),
+      random_num_gen_(RNG_SEED),
+      passing_config_(passing_config)
+{
+}
+
+PassWithRating PassGenerator::getBestPass(const World& world,
+                                          const std::vector<RobotId>& robots_to_ignore)
+{
+    auto receiving_positions_map =
+        sampleReceivingPositionsPerRobot(world, robots_to_ignore);
+
+    // if there are no friendly robots, return early
+    if (receiving_positions_map.empty())
+    {
+        // default pass with 0 rating
+        return PassWithRating{Pass(Point(), Point(), 1.0), 0};
+    }
+
+    // Optimize the receiving positions for each robot and get the best pass
+    PassWithRating best_pass = optimizeReceivingPositions(world, receiving_positions_map);
+
+    // Visualize the sampled passes and the best pass
+    if (passing_config_.pass_gen_vis_config().visualize_sampled_passes())
+    {
+        std::vector<TbotsProto::DebugShapes::DebugShape> debug_shapes;
+        for (const auto& [robot_id, receiving_positions] : receiving_positions_map)
+        {
+            for (const Point& receiving_position : receiving_positions)
+            {
+                debug_shapes.push_back(*createDebugShape(
+                    Stadium(world.friendlyTeam().getRobotById(robot_id)->position(),
+                            receiving_position, 0.02),
+                    std::to_string(debug_shapes.size()) + "gdpg"));
```

Comment:
gdpg?

Reply:
Initially this class was named `GradientDescentPassGenerator` and it stood for that. 
The second argument of `createDebugShape` was updated in #3198 (which I've also merged into this branch) to be a unique ID for the shape being drawn so multiple shapes could have the same text label while being treated as different shapes. So this is an arbitrary name I picked. I've updated to just "pg" now to reflect the current class name